### PR TITLE
Fix Minitest::Spec source location

### LINKED ISSUE #534: Minitest::Spec suites can report production files as test source files
## Summary

When the Ruby tracer instruments `Minitest::Spec`, suites declared with a constant description can report the described production constant file as the test source file instead of the spec file.

The Minitest helper currently resolves the runnable source location from `klass.name` via `Object.const_source_location` before falling back to the test method source location. For specs like `describe Sidekiq::Processor`, the runnable class name resolves to the production constant, so `test.source.file` and test discovery `suiteSourceFile` can point at `lib/sidekiq/processor.rb` instead of `test/processor_test.rb`.

## Example

In Sidekiq Minitest specs:

- `test/processor_test.rb`: `describe Sidekiq::Processor do`
  - reported as `lib/sidekiq/processor.rb`
- `test/middleware_test.rb`: `describe Sidekiq::Middleware do`
  - reported as `lib/sidekiq/middleware/chain.rb`
- `test/web_test.rb`: `describe Sidekiq::Web do`
  - reported as `lib/sidekiq/web/csrf_protection.rb`

These values appeared in ddtest full discovery records as `suiteSourceFile` values, for example:

```json
{"suite":"Sidekiq::Processor at lib/sidekiq/processor.rb","suiteSourceFile":"lib/sidekiq/processor.rb"}
```

## Expected behavior

For `Minitest::Spec`, the source file should identify where the spec/test was declared, e.g. `test/processor_test.rb`, not the implementation file for the described constant.

## Why this matters

Downstream consumers such as ddtest use discovery source files to build runnable test plans and artifacts. If production files are reported as suite source files, those files can be mistaken for runnable test files unless every downstream consumer applies additional test-location filtering.

This also makes source metadata misleading for Test Optimization features that rely on `test.source.file` or suite source file attribution.

## Suspected code path

The relevant Minitest helper appears to be:

- `Datadog::CI::Contrib::Minitest::Helpers.extract_runnable_source_location`
- `Datadog::CI::Contrib::Minitest::Helpers.extract_source_location_from_class`
- `Datadog::CI::SourceCode::ConstantResolver.safely_get_const_source_location`

The problematic order is: resolve `klass.name` with `Object.const_source_location` first, then fall back to `klass.instance_method(method_name).source_location` only when constant resolution is empty.

For `Minitest::Spec` constant descriptions, that first step can resolve the described application constant rather than the test declaration.

### PR DESCRIPTION
**Pull request by anmarchenko**
## Summary

Fix Minitest::Spec suite source resolution when a spec is declared with a constant description, such as `describe SimpleModel`.

For `Minitest::Spec`, the helper now prefers the generated test method source location before falling back to the described class constant source location. This prevents suite names and source attribution from pointing at production files instead of the spec file.

The source-location helpers now normalize empty source locations to `nil`, and the RBS signatures were updated to match the new helper behavior.

Fixes #534.

### CODE REVIEW

File: lib/datadog/ci/contrib/minitest/helpers.rb [hunk 1]
```diff
@@ -81,11 +81,14 @@ def self.test_suite_name(klass, method_name)
...
 
           def self.extract_runnable_source_location(klass, method_name)
-            source_location = extract_source_location_from_class(klass)
-            if source_location.nil? || source_location.empty?
-              return klass.instance_method(method_name).source_location
+            method_source_location = extract_source_location_from_method(klass, method_name)
+            klass_source_location = extract_source_location_from_class(klass)
+
+            if defined?(::Minitest::Spec) && klass&.ancestors&.include?(::Minitest::Spec)
+              method_source_location || klass_source_location
```
[the remark this answers is not in the document]
Comment (anmarchenko) [reply lib/datadog/ci/contrib/minitest/helpers.rb:88]: while possible, this sounds more like an uncommon edge case

**Review by daniel-mohedano — APPROVED** [review]

**Comment by anmarchenko** [comment]
## E2E Test Report: SUCCESS

**Tested by:** Shepherd Agent (autonomous QA for Datadog Test Optimization)

### Test Environment

- **Method:** Local Shepherd/crook testing with the Sidekiq playground and mockdog
- **Playground:** `sidekiq`
- **Focused file:** `test/manager_test.rb` (`describe Sidekiq::Manager`)
- **Revision tested:** `1912c96b69b964087fa59ee390d86e0ec03667c3`
- **Branch tested:** `anmarchenko/fix-minitest-spec-source-location`
- **ddtest:** `main`, built and injected by crook for ddtest validation

### Results

| Check | Status | Evidence |
| --- | --- | --- |
| Baseline reproduces bug on `datadog-ci-rb=main` | PASS | Normal Sidekiq run `20260708-141346` reported suite `Sidekiq::Manager at lib/sidekiq/manager.rb`; ddtest plan `20260708-141453` discovered `suiteSourceFile: lib/sidekiq/manager.rb` and planned `0` files. |
| PR fixes normal Minitest span attribution | PASS | PR run `20260708-140826` passed 3/3 tests; test and suite spans used `test.source.file=test/manager_test.rb` and suite `Sidekiq::Manager at test/manager_test.rb`. |
| PR fixes ddtest full discovery | PASS | PR ddtest plan `20260708-141046` discovered 3 tests / 1 suite with `suiteSourceFile: test/manager_test.rb`; `runner/test-files.txt` contained `test/manager_test.rb`. |
| PR works with ddtest execution | PASS | PR ddtest run `20260708-141144` passed 3/3 tests with ITR enabled; all worker spans used `test.source.file=test/manager_test.rb`. |

### Issues Found

- Confirmed the pre-existing Sidekiq bug on `main`: constant-described Minitest::Spec suites, for example `describe Sidekiq::Manager`, resolve the suite to the production file (`lib/sidekiq/manager.rb`). This breaks ddtest discovery because fast discovery sees `test/manager_test.rb`, while full discovery attaches tests to `lib/sidekiq/manager.rb`, causing the planner to drop the file.
- The PR resolves this: both normal tracing and ddtest discovery/execution now attribute the suite to `test/manager_test.rb`.
- No new issues found.

### Verification

- Local backend payloads were verified with mockdog and `bin/query-spans`.
- Datadog UI verification was not performed because this validation targeted the local mockdog backend, not a real Datadog site.

### Test Methodology

1. Ran a focused Sidekiq Minitest file against `datadog-ci-rb=main` to reproduce the bug.
2. Ran the same focused file against the PR branch with `--debug` and inspected mockdog spans.
3. Ran `ddtest-plan` with `itr-sidekiq-suite-skips` and `DD_TEST_OPTIMIZATION_RUNNER_TESTS_LOCATION=test/manager_test.rb` to force full discovery.
4. Ran full `ddtest` execution for the same scoped file to verify the worker path submits corrected spans.

---

_This E2E test was performed by Shepherd - autonomous QA agent for Datadog Test Optimization._
